### ISSUE TITLE
Problem building 0.3.37-m1 in jetson nano

### ISSUE THREAD

**Issue by DocksDocks** [sskorol/vosk-api-jetson#28]
First I enabled nvidia runtime in my jetson nano,
checked the version of my jetpack:
```
Package: nvidia-jetpack
Version: 4.6.2-b5
Architecture: arm64
Maintainer: NVIDIA Corporation
Installed-Size: 194
Depends: nvidia-cuda (= 4.6.2-b5), nvidia-opencv (= 4.6.2-b5), nvidia-cudnn8 (= 4.6.2-b5), nvidia-tensorrt (= 4.6.2-b5), nvidia-visionworks (= 4.6.2-b5), nvidia-container (= 4.6.2-b5), nvidia-vpi (= 4.6.2-b5), nvidia-l4t-jetson-multimedia-api (>> 32.7-0), nvidia-l4t-jetson-multimedia-api (<< 32.8-0)
Homepage: http://developer.nvidia.com/jetson
Priority: standard
Section: metapackages
Filename: pool/main/n/nvidia-jetpack/nvidia-jetpack_4.6.2-b5_arm64.deb
Size: 29378
SHA256: 925f4abff97e6024d86cff3b9e132e7c7554d05fb83590487381b7e925d5b2bb
SHA1: e3ef727e87df5c331aece34508c110d57d744fe9
MD5sum: 7cb2e387af41bc8143ac7b6525af7794
Description: NVIDIA Jetpack Meta Package
Description-md5: ad1462289bdbc54909ae109d1d32c0a8
```

Added my user to the docker group
```
sudo groupadd docker
sudo usermod -aG docker ${USER}
```
then proceeded to pull the image for arm64:
```
docker pull sskorol/vosk-api:0.3.37-m1
```

Changed the Dockerfile to use the proper tag
```
ARG L4T_VERSION=r32.6.1

FROM sskorol/vosk-api:0.3.37-m1
```

Then proceeded to use the command:
```
cd jetson && ./build.sh -m nano -i ml -t 0.3.37
```

But in step 14 or 15 I had this problem:
It freezes in:
```
c++ -std=c++17 -I.. -isystem /opt/kaldi/tools/openfst-1.8.0/include -O3 -march=armv8-a -Wno-sign-compare -Wall -Wno-sign-comare -Wno-unused-local-typedefs -Wno-deprecated-declarations -Winit-self -DKALDI _DOUBLEPRECISION=0 -DHAVE_EXECINFO_H=1 -DHAVE_CXXABI_H -DHAVE_OPENBLAS -I/opt/kaldi/tools/OpenBLAS/install/include -ftree-vectorize -pthread -g -fPIC -DUSE_KALDI_SVD -c -o word-allign-lattice.o word-allign-lattice.cc
```

After 1 hour or so it comes back and says:
```
c++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
make[1]: *** [<builtin>: lattice-functions-transition-model.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/opt/kaldi/src/lat'
make: *** [Makefile:147: lat] Error 2
The command '/bin/sh -c cd /opt/kaldi/src &&   ./configure --mathlib=OPENBLAS_CLAPACK --shared --use-cuda &&  sed -i "s: -01 : -03 -march=$ARCH :g" kaldi.mk && echo "[BUILDING KALDI] -j $(nproc) online2 lm rnnlm cudafeat cudadecoder" returned a non-zero code: 2
unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /home/jetson/projects/vosk-api-gpu/Dockerfile.server: no such file or directory
```
Im currently using a jetson nano that has 4GB RAM

**Comment by sskorol** [comment]
@DocksDocks I wonder why did you use `m1` image as a baseline for the Jetson Nano? `m1` image is for Apple M1 chip (+ w/o GPU support), and not for Jetson boards. Please read the instruction and find out the links with valid image tags.

**Comment by DocksDocks** [comment]
Thank you so much for the answer, it was actually a lack of attention and knowledge by my part.
I focused on the OS/ARCH part and saw that the m1 was linux/arm64 which is the architecture of my jetson nano, and I thought that it was what I was looking.
Now that you told me to look for valid image tags I saw the [0.3.27-nano] and will try to build it tomorrow. 
I am new to jetson nano and dockers, so I have a question, I was trying to build this to use with python code and basically detect in real time speech with my jetson, is this the correct path?
Thank you so much again for the quick answer and sorry for any inconvenience.

**Comment by DocksDocks** [comment]
@sskorol Thank you so much for the answer, it was actually a lack of attention and knowledge by my part.
I focused on the OS/ARCH part and saw that the m1 was linux/arm64 which is the architecture of my jetson nano, and I thought that it was what I was looking.
Now that you told me to look for valid image tags I saw the [0.3.27-nano] and will try to build it tomorrow.
I am new to jetson nano and dockers, so I have a question, I was trying to build this to use with python code and basically detect in real time speech with my jetson, is this the correct path?
Thank you so much again for the quick answer and sorry for any inconvenience.

**Comment by sskorol** [comment]
@DocksDocks, yes, that's the best way to go. But be aware that 4Gb of memory is not enough for a big (more accurate) model. You can use only a small model on Jetson Nano. I have both Nano and Xavier NX. The latter works with a big model relatively well after making some tweaks.

**Comment by DocksDocks** [comment]
Hello again, this time I tried to build the 0.3.27 nano docker image and at step 15 I got this error this time:
```
XXABI_H -DHAVE_OPENBLAS -I/opt/kaldi/tools/OpenBLAS/install/include -ftree-vectorize -pthread -g -fPIC -DUSE_KALDI_SVD -DHAVE_CUDA -I/usr/local/cuda/include -fPIC -pthread -isystem /opt/kaldi/tools/openfst-1.8.0/include   -c -o nnet-compute-discriminative-parallel.o nnet-compute-discriminative-parallel.cc
c++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
make[1]: *** [nnet-example-functions.o] Error 1
make[1]: *** Waiting for unfinished jobs....
<builtin>: recipe for target 'nnet-example-functions.o' failed
ar -cr kaldi-decoder.a training-graph-compiler.o lattice-simple-decoder.o lattice-faster-decoder.o lattice-faster-online-decoder.o simple-decoder.o faster-decoder.o decoder-wrappers.o grammar-fst.o decodable-matrix.o lattice-incremental-decoder.o lattice-incremental-online-decoder.o
ranlib kaldi-decoder.a
c++ -shared -o libkaldi-decoder.so -Wl,--as-needed  -Wl,-soname=libkaldi-decoder.so,--whole-archive kaldi-decoder.a -Wl,--no-whole-archive  -Wl,-rpath=/opt/kaldi/tools/openfst-1.8.0/lib -rdynamic -Wl,-rpath=/opt/kaldi/src/lib  ../lat/libkaldi-lat.so  ../fstext/libkaldi-fstext.so  ../hmm/libkaldi-hmm.so  ../transform/libkaldi-transform.so  ../gmm/libkaldi-gmm.so  ../tree/libkaldi-tree.so  ../util/libkaldi-util.so  ../matrix/libkaldi-matrix.so  ../base/libkaldi-base.so /opt/kaldi/tools/openfst-1.8.0/lib/libfst.so -L/opt/kaldi/tools/OpenBLAS/install/lib -l:libopenblas.a -l:libblas.a -l:liblapack.a -l:libf2c.a -lm -lpthread -ldl
make[1]: Leaving directory '/opt/kaldi/src/nnet2'
make: *** [nnet2] Error 2
Makefile:147: recipe for target 'nnet2' failed
ln -sf /opt/kaldi/src/decoder/libkaldi-decoder.so /opt/kaldi/src/lib/libkaldi-decoder.so
make[1]: Leaving directory '/opt/kaldi/src/decoder'
The command '/bin/sh -c cd /opt/kaldi/src && 	./configure --mathlib=OPENBLAS_CLAPACK --shared --use-cuda && 	sed -i "s: -O1 : -O3 -march=$ARCH :g" kaldi.mk && 	echo "[BUILDING KALDI] >>>" && 	make -j $(nproc) online2 lm rnnlm cudafeat cudadecoder' returned a non-zero code: 2
unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /home/jetson/projects/vosk-api-gpu/jetson/Dockerfile.server: no such file or directory
```
I thought since I put the right image this time it would work fine, did I do something wrong?

**Comment by sskorol** [comment]
I see 2 issues here:
- it failed on Kaldi compilation
- then it failed to find a server Dockerfile

The latter could be quick fixed by copying https://github.com/sskorol/vosk-api-gpu/blob/main/pc/Dockerfile.server to jetson folder. The former is trickier as the root cause is not clear.

Quick questions:
- did you add SWAP on your Nano?
- did you enable nvidia-docker runtime?

**Comment by DocksDocks** [comment]
@sskorol 
For the questions:
```
did you add SWAP on your Nano?
```
- I don't know what adding swap to my nano means, sorry for being dumb, if you're talking about memory swap i'm using the default jetson nano 4gb config, using 'top' command shows me this:
```
KiB Mem :  4059296 total,  1263096 free,  1662472 used,  1133728 buff/cache
KiB Swap:  2029632 total,  2029632 free,        0 used.  2162484 avail Mem 
``` 
Second question
```
did you enable nvidia-docker runtime?
```
- I did enable my nvidia-docker runtime

And about the file you told me to copy in my jetson folder, after I add the Dockerfile.server to my jetson folder do I need to change the parameters equivalent to my image inside the .server file?
(this is more for confirmation) because it comes with this arg
```
ARG TAG=0.3.37-pc
```
and the $tag
```
FROM sskorol/vosk-api:$TAG
```
Basically I am asking if I need to change the tags like I did in the normal Dockerfile

**Comment by DocksDocks** [comment]
I added the Dockerfile.server and

**Comment by DocksDocks** [comment]
I added the Dockerfile.server to the jetson folder and added the proper tags to the file, increased my SWAP size from 2GB to 6GB (it helped a lot in the proccess) and...

Build is finally completed
```
  ---> 7a30c6d78225
Step 8/8 : CMD ["python3", "./asr_server_gpu.py"]
 ---> Running in c0f36e4b8890
Removing intermediate container c0f36e4b8890
 ---> 6d87c0f051f8
Successfully built 6d87c0f051f8
Successfully tagged sskorol/vosk-server:0.3.27-nano
```

I will open in discussion a thread that's not related to compile so if a person has a problem still uses it, will close this issue now
